code/modules/research/anomaly/anomaly_refinery.dm

```diff
@@ -0,0 +1,300 @@
+	if(active)
+		to_chat(user, span_warning("You can't insert [tool] into [src] while [p_theyre()] currently active."))
+		return
+	if(istype(tool, /obj/item/raw_anomaly_core))
+		if(inserted_core)
+			to_chat(user, span_warning("There is already a core in [src]."))
+			return
+		if(!user.transferItemToLoc(tool, src))
+			to_chat(user, span_warning("[tool] is stuck to your hand."))
+			return
+		var/obj/item/raw_anomaly_core/raw_core = tool
+		if(!get_required_radius(raw_core.anomaly_type))
+			say("Unfortunately, due to diminishing supplies of condensed anomalous matter, [raw_core] and any cores of its type are no longer of a sufficient quality level to be compressed into a working core.")
+		inserted_core = raw_core
+		to_chat(user, span_notice("You insert [raw_core] into [src]."))
+		return
+	if(istype(tool, /obj/item/transfer_valve))
+		if(inserted_bomb)
+			to_chat(user, span_warning("There is already a bomb in [src]."))
+			return
+		var/obj/item/transfer_valve/valve = tool
+		if(!valve.ready())
+			to_chat(user, span_warning("[valve] is incomplete."))
+			return
+		if(!user.transferItemToLoc(tool, src))
+			to_chat(user, span_warning("[tool] is stuck to your hand."))
+			return
+		inserted_bomb = tool
+		tank_to_target = inserted_bomb.tank_two
+		to_chat(user, span_notice("You insert [tool] into [src]"))
+		return
+	if(default_unfasten_wrench(user, tool))
+		return
+	if(default_deconstruction_screwdriver(user, "[base_icon_state]-off", "[base_icon_state]", tool))
+		return
+	if(default_deconstruction_crowbar(tool))
+		return
+	update_appearance()
+	return ..()
+
+/**
+ * Starts a compression test.
+ */
+/obj/machinery/research/anomaly_refinery/proc/start_test()
+	if(!istype(inserted_core) || !istype(inserted_bomb))
+		end_test("ERROR: Missing equpment. Items ejected.")
+		return
+
+	if(!inserted_bomb?.tank_one || !inserted_bomb?.tank_two || !(tank_to_target == inserted_bomb?.tank_one || tank_to_target == inserted_bomb?.tank_two))
+		end_test("ERROR: Transfer valve malfunctioning. Items ejected.")
+		return
+
+	say("Beginning compression test. Opening transfer valve.")
+	active = TRUE
+	test_status = null
+	inserted_bomb.toggle_valve(tank_to_target)
+	tank_to_target = null
+	timeout_timer = addtimer(CALLBACK(src, .proc/timeout_test), COMPRESSION_TEST_TIME, TIMER_STOPPABLE | TIMER_UNIQUE | TIMER_NO_HASH_WAIT)
+	return
+
+/**
+ * Ends a compression test.
+ *
+ * Arguments:
+ * - message: A message for the compressor to say when the test ends.
+ */
+/obj/machinery/research/anomaly_refinery/proc/end_test(message)
+	active = FALSE
+	if(inserted_core)
+		eject_core()
+	if(inserted_bomb)
+		eject_bomb()
+	if(timeout_timer)
+		QDEL_NULL(timeout_timer)
+	if(message)
+		say(message)
+	return
+
+/**
+ * Checks whether an internal explosion was sufficient to compress the core.
+ */
+/obj/machinery/research/anomaly_refinery/proc/check_test(atom/source, list/arguments)
+	SIGNAL_HANDLER
+	. = COMSIG_CANCEL_EXPLOSION
```

Comment:
why?

Reply:
Not my code. It's to cancel the explosion using the return signal. https://github.com/tgstation/tgstation/blob/11ccf197416a73bd3e187d7a035529e6112861af/code/controllers/subsystem/explosions.dm#L244

Reply:
Honestly it makes tracking the flow of the program much harder. People shouldn't need to keep that sort of thing in their head when reading a proc
See <https://github.com/tgstation/tgstation/pull/64940/files>

Reply:
oh dear thats new